### PR TITLE
functors task (2) - Pysarenko

### PR DESCRIPTION
**Pull request by Elzianor**

### CODE REVIEW

File: functors_task_2/student/include/impl.h [hunk 1]
```diff
@@ -62,10 +62,10 @@ class DataBrowser
...
     {
         static_assert(std::is_same<decltype(f(nullptr)), bool>::value, "Provided Callable must return bool");
-        // find reader
-        // check for errors
-        // call functor
-        return false;
+
+        const auto& it = m_dataReaders.find(userId);
+
+        return m_dataReaders.cend() == it || !it->second ? false : f(it->second);
```
Comment (smay1613) [inline functors_task_2/student/include/impl.h:68]: too much for one line

File: functors_task_2/student/src/impl.cpp [hunk 2]
```diff
@@ -7,48 +7,44 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    return it->second->getDataType1(returnValues, 0);
+    return safeCall(userId, [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        return invokeDataRequest([](const IDataSelector* selector, std::vector<size_t>& returnValues) {
+            return selector->getDataType1(returnValues, 0);
+        },
+        selector,
+        returnValues);
+    });
+
+    return true;
```
Comment (smay1613) [inline functors_task_2/student/src/impl.cpp:18]: double return?

File: functors_task_2/student/src/impl.cpp [hunk 3]
```diff
@@ -7,48 +7,44 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    {
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType1(returnValues, 0);
+    return safeCall(userId, [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        return invokeDataRequest([](const IDataSelector* selector, std::vector<size_t>& returnValues) {
```
Comment (smay1613) [inline functors_task_2/student/src/impl.cpp:11]: please name that lambda

File: functors_task_2/student/src/impl.cpp [hunk 4]
```diff
@@ -7,48 +7,44 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    if (it == m_dataReaders.cend())
-    {
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType2(returnValues);
+    return safeCall(userId, [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) -> bool {
```
Comment (smay1613) [inline functors_task_2/student/src/impl.cpp:23]: you shouldn't specify return type of lambda here

File: functors_task_2/student/src/impl.cpp [hunk 5]
```diff
@@ -7,48 +7,44 @@ void DataBrowser::userLeave(const std::string &userId)
...
-        return false;
-    }
-
-    std::deque<size_t> unprocessedResults {};
-    bool success {it->second->getDataType3(unprocessedResults)};
-    returnValues = process(unprocessedResults);
-    return success;
+    return safeCall(userId, [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        std::deque<size_t> unprocessedResults {};
+        bool success {invokeDataRequest([](const IDataSelector* selector, std::deque<size_t>& returnValues) {
```
Comment (smay1613) [inline functors_task_2/student/src/impl.cpp:38]: please name this lambda, hard to read

File: functors_task_2/student/include/impl.h (hunk 1 above)
Comment (Elzianor) [reply functors_task_2/student/include/impl.h]: Splitted

File: functors_task_2/student/src/impl.cpp (hunk 2 above)
Comment (Elzianor) [reply functors_task_2/student/src/impl.cpp]: Fixed, second return removed

File: functors_task_2/student/src/impl.cpp (hunk 3 above)
Comment (Elzianor) [reply functors_task_2/student/src/impl.cpp]: Done

File: functors_task_2/student/src/impl.cpp (hunk 4 above)
Comment (Elzianor) [reply functors_task_2/student/src/impl.cpp]: Type removed

File: functors_task_2/student/src/impl.cpp (hunk 5 above)
Comment (Elzianor) [reply functors_task_2/student/src/impl.cpp]: Done

File: functors_task_2/student/include/impl.h [hunk 6]
```diff
@@ -65,7 +65,14 @@ class DataBrowser
 
         const auto& it = m_dataReaders.find(userId);
 
-        return m_dataReaders.cend() == it || !it->second ? false : f(it->second);
+        bool result = false;
```
Comment (smay1613) [inline functors_task_2/student/include/impl.h:68]: please avoid one return rule
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nr2-dont-insist-to-have-only-a-single-return-statement-in-a-function

**Pull request closed without merging**